extensions/stackdriver/stackdriver.cc

```diff
@@ -143,32 +144,44 @@ void StackdriverContext::onLog() {
 
   // Fill in peer node metadata in request info.
   if (isOutbound) {
-    google::protobuf::Struct upstream_metadata;
-    if (getMetadataStruct(Common::Wasm::MetadataType::Request,
-                          kUpstreamMetadataKey,
-                          &upstream_metadata) != Common::Wasm::WasmResult::Ok) {
-      logWarn(absl::StrCat("cannot get metadata for: ", kUpstreamMetadataKey));
+    auto metadata_value =
+        getSelectorExpression({"filter_state", kUpstreamMetadataKey});
+    if (!metadata_value) {
+      logWarn(absl::StrCat("cannot get stackdriver metadata for: ",
+                           kUpstreamMetadataKey));
       return;
     }
-
-    auto status = ::Wasm::Common::extractNodeMetadata(upstream_metadata,
-                                                      &peer_node_info_);
+    google::protobuf::Value upstream_metadata;
+    if (!upstream_metadata.ParseFromArray(metadata_value.value()->data(),
+                                          metadata_value.value()->size())) {
+      logWarn(absl::StrCat("cannot parse stackdriver metadata for: ",
+                           kUpstreamMetadataKey));
+      return;
+    }
+    auto status = ::Wasm::Common::extractNodeMetadata(
+        upstream_metadata.struct_value(), &peer_node_info_);
     if (status != Status::OK) {
       logWarn("cannot parse upstream peer node metadata " +
               upstream_metadata.DebugString() + ": " + status.ToString());
     }
   } else {
-    google::protobuf::Struct downstream_metadata;
-    if (getMetadataStruct(Common::Wasm::MetadataType::Request,
-                          kDownstreamMetadataKey, &downstream_metadata) !=
-        Common::Wasm::WasmResult::Ok) {
-      logWarn(
-          absl::StrCat("cannot get metadata for: ", kDownstreamMetadataKey));
+    auto metadata_value =
+        getSelectorExpression({"filter_state", kDownstreamMetadataKey});
+    if (!metadata_value) {
+      logWarn(absl::StrCat("cannot get stackdriver metadata for: ",
+                           kDownstreamMetadataKey));
+      return;
+    }
+    google::protobuf::Value downstream_metadata;
```

Comment:
Can this be parsed into google::protobuf::Struct directly instead of google::protobuf::Value?

Reply:
The ABI function underlying it produces Value instead of Struct since it has to support lists and other values. So the output is the generic Value but then we know that it contains a struct. The overhead is not that big, just a few extra bytes.